### PR TITLE
Add salt and dependencies

### PR DESCRIPTION
**Pull request by danielfrg**
Adds [salt](https://github.com/saltstack/salt) (Linux and OS X and Py2) and the remaining dependencies: [apache-libcloud](https://libcloud.apache.org/) (Win, OSX, Linux Py2 and Py3)

### CODE REVIEW

File: recipes/apache-libcloud/meta.yaml [hunk 1]
```diff
@@ -0,0 +1,33 @@
...
+  name: apache-libcloud
+  version: {{ version }}
+
+source:
+  fn: apache-libcloud-{{ version }}.tar.bz2
+  url: http://apache.osuosl.org/libcloud/apache-libcloud-{{ version }}.tar.bz2
+  sha1: fbf558d7e52336042dd9a456dec3a73f4b35e38e
+
+build:
+  number: 0
```
Comment (janschulz) [inline recipes/apache-libcloud/meta.yaml]: You can use the following here instead of `bld.bat` and `build.sh`
```
build:
[...]
  script: python setup.py install --single-version-externally-managed --record record.txt
```

Comment (danielfrg) [inline recipes/apache-libcloud/meta.yaml]: Thanks, I guess I am used to the scripts :). I change it.

File: recipes/salt/meta.yaml [hunk 2]
```diff
@@ -0,0 +1,43 @@
+{% set version = "2015.8.7" %}
+
+package:
+  name: salt
+  version: {{ version }}
+
+source:
+  git_tag: v{{ version }}
+  git_url: https://github.com/saltstack/salt.git
```
Comment (jakirkham) [inline recipes/salt/meta.yaml]: Can we please use a tarball instead?

Comment (ocefpaf) [inline recipes/salt/meta.yaml]: @danielfrg you can use these 3 lines instead of the above:

```yaml
fn: salt-{{ version }}.tar.gz
url: https://github.com/saltstack/salt/archive/v{{ version }}.tar.gz
sha256: 5cb7460741f44db445baf56e88acb5921e43dbefd5a8c24496f9b78718d96094
```

**Comment by danielfrg** [comment]
Ok, I give up on the windows build for salt right now. I'll keep working on it with the remote connection in AppVeyor once the feedstock is created.

I made the changes to use the tarbal and remove the unnecessary build scripts. Let me know if there is something else blocking this.

**Comment by danielfrg** [comment]
A hack here and there and I was able to make salt work on windows. We should be good to go with this PR unless anybody has any comments.

File: recipes/salt/bld.bat [hunk 3]
```diff
@@ -0,0 +1,18 @@
+%PYTHON% -c "import os; import sys; import urllib; urllib.urlretrieve('http://repo.saltstack.com/windows/dependencies/64/libeay32.dll', filename=os.path.join(os.path.dirname(sys.executable), 'libeay32.dll'))"
```
Comment (patricksnape) [inline recipes/salt/bld.bat]: These should exist in OpenSSL?

File: recipes/salt/meta.yaml [hunk 4]
```diff
@@ -0,0 +1,46 @@
...
+  fn: salt-{{ version }}.tar.gz
+  url: https://github.com/saltstack/salt/archive/v{{ version }}.tar.gz
+  sha256: 93670d3a280e4a4d32fa12e2d5aaaa2ac7cb309c8f7fafff4dd8130fba7b40ff
+
+build:
+  number: 0
+  skip: True  # [win32 or py3k]
+
+requirements:
+  build:
```
Comment (patricksnape) [inline recipes/salt/meta.yaml]: I think if you add OpenSSL on Windows you won't need those copy hacks.

Comment (danielfrg) [inline recipes/salt/meta.yaml]: `m2crypto` requires `openssl` so it already a dependency. I can add it here to make it explicit.

Those dll file copy/downloads are required by the salt `setup.py` (https://github.com/saltstack/salt/blob/develop/setup.py#L418-L427) but for some reason that code is not working in appveyor so I downloaded them manually before calling `setup.py`.

Comment (patricksnape) [inline recipes/salt/meta.yaml]: But they should be on the path so they shouldn't be required. They should be there via openssl. Since the `setup.py` is forcing this it feels like that is incorrect and a patch should be made upstream to allow this to be optional - since we satisfy that dependency ourselves. If you copy the files there then they will get shipped as part of the recipe and thus we are polluting the root of the conda environment with these two files.

Comment (danielfrg) [inline recipes/salt/meta.yaml]: Yes, I agree that its not ideal. I can start a discussion upstream to update the `setup.py`. Right now those options are basically hardcoded if salt is being installed on windows.

I think that the easiest way for now would be to have a patch on the `setup.py` that its applied on build.

Comment (danielfrg) [inline recipes/salt/meta.yaml]: I made a small patch on the `setup.py` to not download those dll files and everything seams to be working. Thanks for the feedback!

**Comment by danielfrg** [comment]
Anything else I have to do here @conda-forge/core?

File: recipes/salt/meta.yaml (hunk 4 above)
Comment (jakirkham) [inline recipes/salt/meta.yaml]: We should open an issue upstream explaining this issue and asking for some option to get around it.

Comment (danielfrg) [inline recipes/salt/meta.yaml]: I have created an issue here: https://github.com/saltstack/salt/issues/33848 - I will try to make a PR to fix this upstream when I have some time.

**Comment by jakirkham** [comment]
LGTM

**Comment by jakirkham** [comment]
@patricksnape, did you have more thoughts here or are you happy?

**Comment by patricksnape** [comment]
LGTM - thanks for the hard work @danielfrg!

**Comment by danielfrg** [comment]
Thanks for the comments and help. I am excited about this one specially for windows, even tho I don't use it on windows :P

File: recipes/salt/build.sh [hunk 5]
```diff
@@ -0,0 +1,30 @@
+#!/bin/bash
+
+SRC_DIR="."
```
Comment (ocefpaf) [inline recipes/salt/build.sh]: `conda-build` already defines `SRC_DIR` for you.

File: recipes/salt/build.sh [hunk 6]
```diff
@@ -0,0 +1,30 @@
...
+var/run/salt
+srv/salt
+srv/pillar
+var/log/salt
+var/run
+"
+for path in $DIRECTORIES
+do
+    mkdir -p $PREFIX/$path
+    touch $PREFIX/$path/.gitkeep
```
Comment (ocefpaf) [inline recipes/salt/build.sh]: Just curious. Why is this needed?

Comment (danielfrg) [inline recipes/salt/build.sh]: Salt uses that structure of directories to store some information (keys and configs) i think some of these are optional but i found it better to create them in the recipe.

File: recipes/salt/build.sh (hunk 5 above)
Comment (danielfrg) [inline recipes/salt/build.sh]: Thanks, i removed it.

File: recipes/salt/build.sh (hunk 6 above)
Comment (ocefpaf) [inline recipes/salt/build.sh]: Got it. But why the `.gitkeep`?

**Comment by jakirkham** [comment]
Sorry on phone so diff commenting is not working.

As those directories are empty otherwise, I believe `conda-build` will drop them when creating the package. The stub files make sure they are kept. The naming of the stub files is arbitrary. Does this correctly summarize the situation, @danielfrg?

**Comment by danielfrg** [comment]
Yes, thats 100% correct. I can change the `.gitkeep` name - its something I have used in the past for `git`.

**Comment by jakirkham** [comment]
Sure. How about we call `.conda_keep`? I know it is kind of a tedious change, but `.gitkeep` makes sound like it relates to `git` even though it doesn't. This at least makes the intent a little more clear so that people know what we are trying to accomplish.

Maybe in the future `conda-build` could be designed to clean up these empty files and save the directories instead. @msarahan, any thoughts on this last idea? I probably could write a very simple PR to add this functionality if it would be acceptable, but by no means is this high priority just interesting.

**Comment by danielfrg** [comment]
Its a fair change i think I renamed it from `.gitkeep` to `.condakeep`.

**Comment by msarahan** [comment]
Could Conda build just keep empty directories? Seems roundabout to have files that are only removed.

**Comment by jakirkham** [comment]
> Could Conda build just keep empty directories?

I don't believe it does, but I have not tried recently. I'm guessing @danielfrg has checked recently though.

> Seems roundabout to have files that are only removed. 

Sure, as it does with `git` too. Still the workaround is so trivial it is not a problem.

If we can just keep empty directories, that would be preferred. 👍

**Comment by danielfrg** [comment]
I checked a while back and i dont think this have changed.

**Comment by msarahan** [comment]
https://github.com/conda/conda-build/issues/1014

**Comment by jakirkham** [comment]
Thanks for opening that issue @msarahan. Is there another you would rather have us address this for now or is the current solution an acceptable workaround until this is addressed in `conda`/`conda-build`?

**Comment by msarahan** [comment]
Current work around is totally fine with me.

**Comment by danielfrg** [comment]
Cool, are we good to merge this one?

**Comment by msarahan** [comment]
Thanks!

**Comment by jakirkham** [comment]
@thatch45 @s0undt3ch @terminalmage @nmadhok @cachedout @rallytime @techhat @whiteinge @jacksontj @basepi, just letting you know that we have added `salt` to [conda-forge]( http://conda-forge.github.io/ ). This will allow users to easily install this with [`conda`]( http://conda.pydata.org/docs/ ) ( [`conda install -c conda-forge salt`]( http://anaconda.org/conda-forge/salt ) ). The maintenance burden here is light as all releases are built and deployed automatically with CI. Changes to how it is being built can be controlled by changing the recipe in [this feedstock (repo)]( https://github.com/conda-forge/salt-feedstock ). If any of you have any interest in being added as a maintainer (useful for making new releases for instance), we would be more than happy to add you to the maintainers list.

**Comment by jakirkham** [comment]
@Kami @tonybaloney @erjohnso, just letting you know that we have added `libcloud` as `apache-libcloud` to [conda-forge]( http://conda-forge.github.io/ ). This will allow users to easily install this with [`conda`]( http://conda.pydata.org/docs/ ) ( [`conda install -c conda-forge apache-libcloud`]( http://anaconda.org/conda-forge/apache-libcloud ) ). The maintenance burden here is light as all releases are built and deployed automatically with CI. Changes to how it is being built can be controlled by changing the recipe in [this feedstock (repo)]( https://github.com/conda-forge/apache-libcloud-feedstock ). If any of you have any interest in being added as a maintainer (useful for making new releases for instance), we would be more than happy to add you to the maintainers list.

**Comment by tonybaloney** [comment]
@jakirkham @danielfrg note apache-libcloud uses httplib to connect to the cloud APIs, so SSL validation is done using a common list of CA repositories. This requires

* openssl package on CentOS / Fedora (normally there)
* ca-certificates package on Debian / Ubuntu / Arch / Gentoo (normally there)
* ca_root_nss port on FreeBSD
* curl-ca-bundle port on Mac OS X
* openssl and curl-ca-bundle homebrew package

**Comment by jakirkham** [comment]
Thanks for the note, @tonybaloney. We currently supply certificates based on [`certifi`]( https://pypi.python.org/pypi/certifi ).

**Comment by tonybaloney** [comment]
In which case @jakirkham , users will need to run:
```python
import libcloud.security
import certifi
libcloud.security.CA_CERTS_PATH.append(certifi.where())
```
To get the cert validation. I'll have a think about whether libcloud should discover `certifi` package in the site-packages and automatically add the path. It would certainly be easier to use.

**Comment by jakirkham** [comment]
Sorry when I said "based on" I meant we repackage the `cacert.pem` file from and place it in `$PREFIX/ssl`. Though using `certifi` is still an option for end-users. Is there some way we should be telling `libcloud` to look there automatically?

**Comment by jakirkham** [comment]
Ok, I proposed some changes that should address this, @tonybaloney, in this PR ( https://github.com/conda-forge/apache-libcloud-feedstock/pull/1 ). It is pretty hacky, but I merely wanted to get the ball rolling for @danielfrg. It seems to work, but maybe we can work together to come up with something a bit nicer that would be worthy of a PR upstream as well.

**Comment by nmadhok** [comment]
@jakirkham This is awesome stuff! I would definitely love to be added as one of the maintainers and help however I can.

**Comment by basepi** [comment]
@jakirkham please feel free to add me as a maintainer as well. :)

**Comment by jakirkham** [comment]
Adding you guys in this PR ( https://github.com/conda-forge/salt-feedstock/pull/2 ), @nmadhok @basepi.

**Comment by Kami** [comment]
Awesome, great to hear that :+1: 

As far as certificate paths go - one thing we could do is to automatically add certifi CA bundle path to Liblcoud CA bundle path if certifi is installed and available.

Another option, if `%PREFIX%` is somewhat static, would be for us to directly add those paths to Libcloud code base. This way you don't need to maintain a patch...

**Comment by thatch45** [comment]
Awesome!

**Comment by Kami** [comment]
Here is the PR for our certifi related change - https://github.com/apache/libcloud/pull/812. Feedback is welcome.

Also, this PR has been merged, does this mean Libcloud is now available in conda? Just wanted to confirm so I can update "Installation Methods" section on our website.

**Comment by danielfrg** [comment]
It is available in the conda-forge channel not the main conda channel. You can install it like: `conda install -c conda-forge apache-libcloud`

**Comment by Kami** [comment]
@danielfrg Great, thanks.

Please also keep us posted once it's available in the main channel so we can update our website with that information then as well :)
